File: gma-next-gen-sdk-android-examples-main/java/nextgenexample/app/src/main/java/com/example/nextgenexample/banner/BannerFragment.java
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.nextgenexample.banner;

import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.example.nextgenexample.AdFragment;
import com.example.nextgenexample.Constant;
import com.example.nextgenexample.databinding.FragmentBannerBinding;
import com.google.android.libraries.ads.mobile.sdk.banner.AdSize;
import com.google.android.libraries.ads.mobile.sdk.banner.BannerAd;
import com.google.android.libraries.ads.mobile.sdk.banner.BannerAdEventCallback;
import com.google.android.libraries.ads.mobile.sdk.banner.BannerAdRefreshCallback;
import com.google.android.libraries.ads.mobile.sdk.banner.BannerAdRequest;
import com.google.android.libraries.ads.mobile.sdk.common.AdLoadCallback;
import com.google.android.libraries.ads.mobile.sdk.common.LoadAdError;

/** An [AdFragment] subclass that loads a banner ad. */
public class BannerFragment extends AdFragment<FragmentBannerBinding> {
  // Default constructor required for fragment instantiation.
  public BannerFragment() {}

  // Sample banner ad unit ID.
  private static final String AD_UNIT_ID = "ca-app-pub-3940256099942544/9214589741";
  private BannerAd bannerAd;
  private FrameLayout bannerViewContainer;

  @Override
  protected BindingInflater<FragmentBannerBinding> getBindingInflater() {
    return FragmentBannerBinding::inflate;
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);

    // Get the ad size based on the screen width.
    AdSize adSize =
        AdSize.getCurrentOrientationAnchoredAdaptiveBannerAdSize(requireContext(), getAdWidth());

    // Give the banner container a placeholder height to avoid a sudden change
    // when the ad loads.
    bannerViewContainer = binding.bannerViewContainer;
    ViewGroup.LayoutParams bannerLayoutParams = bannerViewContainer.getLayoutParams();
    bannerLayoutParams.height = adSize.getHeightInPixels(requireContext());
    bannerViewContainer.setLayoutParams(bannerLayoutParams);

    // Load an ad.
    loadAd(adSize);
  }

  // Determine the screen width to use for the ad width.
  private int getAdWidth() {
    DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
    int adWidthPixels = displayMetrics.widthPixels;

    // The Next Gen SDK uses DisplayMetrics to check screen width. Avoid using window manager bounds
    // on Android R+.
    // if (VERSION.SDK_INT >= VERSION_CODES.R && getActivity() != null) {
    //  WindowMetrics windowMetrics = getActivity().getWindowManager().getCurrentWindowMetrics();
    //  adWidthPixels = windowMetrics.getBounds().width();
    // }

    float density = displayMetrics.density;
    return (int) (adWidthPixels / density);
  }

  private void loadAd(AdSize adSize) {
    if (bannerAd != null) {
      Log.d(Constant.TAG, "Banner ad already loaded.");
      return;
    }

    BannerAd.load(
        new BannerAdRequest.Builder(AD_UNIT_ID, adSize).build(),
        new AdLoadCallback<BannerAd>() {
          @Override
          public void onAdLoaded(@NonNull BannerAd bannerAd) {
            bannerAd.setAdEventCallback(
                new BannerAdEventCallback() {
                  @Override
                  public void onAdImpression() {
                    Log.d(Constant.TAG, "Banner ad recorded an impression.");
                  }

                  @Override
                  public void onAdClicked() {
                    Log.d(Constant.TAG, "Banner ad clicked.");
                  }
                });

            bannerAd.setBannerAdRefreshCallback(
                new BannerAdRefreshCallback() {
                  @Override
                  public void onAdRefreshed() {
                    showToast("Banner ad refreshed.");
                    Log.d(Constant.TAG, "Banner ad refreshed.");
                  }

                  @Override
                  public void onAdFailedToRefresh(@NonNull LoadAdError adError) {
                    showToast("Banner ad failed to refresh with error code: " + adError.getCode());
                    Log.w(Constant.TAG, "Banner ad failed to refresh: " + adError);
                  }
                });

            BannerFragment.this.bannerAd = bannerAd;
            if (getActivity() != null) {
              getActivity()
                  .runOnUiThread(
                      () -> bannerViewContainer.addView(bannerAd.getView(getActivity())));
            }
          }

          @Override
          public void onAdFailedToLoad(@NonNull LoadAdError adError) {
            bannerAd = null;
            showToast("Banner ad failed to load with error code: " + adError.getCode());
            Log.w(Constant.TAG, "Banner ad failed to load: " + adError);
          }
        });
  }
}

File: gma-next-gen-sdk-android-examples-main/kotlin/nextgenexample/app/src/main/java/com/example/next_gen_example/banner/BannerFragment.kt
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.next_gen_example.banner

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.fragment.app.Fragment
import com.example.next_gen_example.AdFragment
import com.example.next_gen_example.Constant
import com.example.next_gen_example.databinding.FragmentBannerBinding
import com.google.android.libraries.ads.mobile.sdk.banner.AdSize
import com.google.android.libraries.ads.mobile.sdk.banner.BannerAd
import com.google.android.libraries.ads.mobile.sdk.banner.BannerAdEventCallback
import com.google.android.libraries.ads.mobile.sdk.banner.BannerAdRefreshCallback
import com.google.android.libraries.ads.mobile.sdk.banner.BannerAdRequest
import com.google.android.libraries.ads.mobile.sdk.common.AdLoadCallback
import com.google.android.libraries.ads.mobile.sdk.common.LoadAdError

/** A simple [Fragment] subclass that loads a banner ad. */
class BannerFragment : AdFragment<FragmentBannerBinding>() {
  override val bindingInflater: (LayoutInflater, ViewGroup?, Boolean) -> FragmentBannerBinding
    get() = FragmentBannerBinding::inflate

  private var bannerAd: BannerAd? = null
  private lateinit var bannerContainer: FrameLayout

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    // Get the ad size based on the screen width.
    val adSize = AdSize.getCurrentOrientationAnchoredAdaptiveBannerAdSize(requireContext(), adWidth)

    // Give the banner container a placeholder height to avoid a sudden change
    // when the ad loads.
    bannerContainer = binding.bannerViewContainer
    val bannerLayoutParams = bannerContainer.layoutParams
    bannerLayoutParams.height = adSize.getHeightInPixels(requireContext())
    bannerContainer.layoutParams = bannerLayoutParams

    // Load an ad.
    loadAd(adSize)
  }

  // Determine the screen width to use for the ad width.
  private val adWidth: Int
    get() {
      val displayMetrics = resources.displayMetrics
      val adWidthPixels = displayMetrics.widthPixels
      val density = displayMetrics.density
      return (adWidthPixels / density).toInt()
    }

  private fun loadAd(adSize: AdSize) {
    if (bannerAd != null) {
      Log.d(Constant.TAG, "Banner ad already loaded.")
      return
    }

    BannerAd.load(
      BannerAdRequest.Builder(AD_UNIT_ID, adSize).build(),
      object : AdLoadCallback<BannerAd> {
        override fun onAdLoaded(ad: BannerAd) {
          ad.adEventCallback =
            object : BannerAdEventCallback {
              override fun onAdImpression() {
                Log.d(Constant.TAG, "Banner ad recorded an impression.")
              }

              override fun onAdClicked() {
                Log.d(Constant.TAG, "Banner ad recorded a click.")
              }
            }
          ad.bannerAdRefreshCallback =
            object : BannerAdRefreshCallback {
              override fun onAdRefreshed() {
                showToast("Banner ad refreshed.")
                Log.d(Constant.TAG, "Banner ad refreshed.")
              }

              override fun onAdFailedToRefresh(adError: LoadAdError) {
                showToast("Banner ad failed to refresh.")
                Log.w(Constant.TAG, "Banner ad failed to refresh: $adError")
              }
            }

          bannerAd = ad
          activity?.runOnUiThread {
            bannerContainer.removeAllViews()
            bannerContainer.addView(ad.getView(requireActivity()))
          }

          showToast("Banner ad loaded.")
          Log.d(Constant.TAG, "Banner ad loaded.")
        }

        override fun onAdFailedToLoad(adError: LoadAdError) {
          bannerAd = null
          showToast("Banner failed to load.")
          Log.w(Constant.TAG, "Banner ad failed to load: $adError")
        }
      },
    )
  }

  companion object {
    // Sample anchored adaptive banner ad unit ID.
    private const val AD_UNIT_ID = "ca-app-pub-3940256099942544/9214589741"
  }
}
